content/systems/filecoin_nodes/node_types/_index.md

```diff
@@ -6,4 +6,22 @@ dashboardWeight: 1
+---
+
+Two types of nodes: full node and miner node. Full (validation) node is a repository, interacts with miner. Full nodes read the chain. Full node needs to reach consensus and do chain sync.
+
+Miner creates nodes. A different repository. Does full validation, extends the chain.
+
+Fundamentally, there are only two major node types in the Filecoin network and this is how the Lotus implementation is realising _node types_ in the strict sense of the word. These are:
+
+The **Full Node:** this is the chain validation or chain verifier node. In essense, this type of node is a repository that interacts with one or more miner node(s). A Full Node must synchronise the chain (ChainSync) when it first joins the network. From then on, the node must read the chain and reach consensus state.
+The **Miner Node:** Miner nodes are the backbone of the Filecoin blockchain network. They have all the functionality of the Full Node, but also extend this functionality to create blocks, submit blocks to the blockchain and ultimately extend the blockchain.
+
+In order to realise the full potential of the Filecoin network, there is more functionality that needs to be added on top of the _base_ node types described above. Different node types are not mutually exclusive to each other, meaning that one physical node might have to implement more than one type of node functionality.
+
+A Filecoin implementation should support the following types of nodes:
+
+- **Chain Verifier Node:** this is the **Full Node** described above. This type of node cannot play an active role in the network, unless it implements **Client Node** functionality, described below.
+- **Client Node:** this type of node builds on top of the **Full or Chain Verifier Node** and must be implemented by any application that is building on the Filecoin network. This can be thought of as the main infrastructure node (at least as far as interaction with the blockchain is concerned) of applications such as exchanges or decentralised storage applications building on Filecoin. The node should implement and interact (as a client) with the Storage and Retrieval Markets, keep the Market Order Book and be able to do Data Transfers through the Data Transfer Module.
+- **Retrieval Miner Node:** this node type is extending the **Full or Chain Verifier Node** to add _retrieval miner_ functionality, that is, participate in the retrieval market. As such, this node type needs to implement the retrieval market provider subsystem, keep the Market Order Book and be able to do Data Transfers through the Data Transfer Module.
+- **Storage Miner Node:** this type of node is building on top of the Miner Node described above and must implement all of its functionality for validating, creating and adding blocks to the blockchain. It should implement the storage mining subsystem, the storage market provider subsystem, keep the Market Order Book and be able to do Data Transfers through the Data Transfer Module.
```

Comment:
Ok, now I think I follow. What you're calling nodes here are what the old spec refers to as services or subsystems. There are only two nodes, full and miner, the rest are services of which the nodes are composed. In fact, a node can be seen as nothing more than the API (services) it exposes, there is nothing intrinsically different in the full node apart from the fact that it's running the sync manager. (Maybe I'm misreading this and the retrieval/storage miner nodes are conceptual abstractions used here in the spec.)

Reply:
>Most importantly, there's nothing to prevent an implementation from having all those as totally separate node types, right?

Theoretically I think no, not sure if it has been tried out.

Reply:
@yiannisbot the distinction I would make is this:
- services can be provided by a node. some services include:
   - chain verification
   - storage market client
   - storage market provider
   - retrieval market client
   - retrieval market provider
   - storage mining
- a node is combination of these services that can participate as a node in the network. I think it's important that a node is ALWAYS someone who participates in the network, and MUST implement the chain verification service as a baseline in order to be considered a node.
- a process is just an OS process, and in our context, it packages a set of services together (you can call it a subsystem, though I find that term confusing) . A process may be a node on its own, or not -- it has to be on the network as a chain verifier to be considered a node.
- a "type of node" describes a common combination of services to make a filecoin node that plays a given role on the network. It can be a single OS process or multiple OS processes. I think the above descriptions (verifier, client, retrieval miner, storage miner) are accurate and should be left as is. These are all the intended types of nodes that will likely exist on the network at some point.

- Where I think you are actually getting lost it trying to map this on to lotus processes.
- When we say "full node" in Lotus, we're referring to the `lotus daemon` process. You're saying a "full node" is a chain verification node. However, running the `Lotus daemon` process on its own is actually a client node:
   - it has the chain verification system + message publishing functionality
   - it has the client services (storage market client & retrieval market client)
   - it has data transfer
- I find the term Full Node confusing and I do not think it should be in the spec
- the `lotus-storage-miner` process is:
   - the storage mining subsystem
   - the storage market provider
   - the retrieval market provider
   - data transfer for the provider
- the `lotus-storage-miner` process, on it's own, is not a node. In fact, it won't do much run on it's own.
- Running BOTH the `lotus daemon` and `lotus-storage-miner` process, absent customization, is both a Storage Mining Node, a Retrieval Mining Node, and a Client Node
  - it has the chain verification system + message publishing functionality
  - it has the client services as well (though they're not often used)
  - it has the markets provider services (both retrieval & storage) and the storage mining subsystem to seal pieces and min blocks
  - it has data transfer for client side and provider side

FWIW, there are other nodes coming:
- The initial Forrest implementation is a "verifier node" -- it has no client services at all, just chain verification
- however, the forrest process can be connected (or will be able to be connected) to `lotus-storage-miner` to make a Storage Miner Node and a Retrieval Miner Node
- the forrest team is making a go process which runs just the client subsystems -- that will together with Forrest make a Client Node
- I assume at some point someone is going to extract only the retrieval parts of the `lotus-storage-miner` process and back it with a purely unsealed data storage mechanism to make a process that when combined with `lotus` or Forrest will make a true "Retrieval Miner Node" who is not also a "Storage Miner Node"